src/simulators/statevector/chunk/chunk_container.hpp

```diff
@@ -0,0 +1,512 @@
+/**
+ * This code is part of Qiskit.
+ *
+ * (C) Copyright IBM 2018, 2019, 2020.
+ *
+ * This code is licensed under the Apache License, Version 2.0. You may
+ * obtain a copy of this license in the LICENSE.txt file in the root directory
+ * of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Any modifications or derivative works of this code must retain this
+ * copyright notice, and modified files need to carry a notice indicating
+ * that they have been altered from the originals.
+ */
+
+
+#ifndef _qv_chunk_container_hpp_
+#define _qv_chunk_container_hpp_
+
+#ifdef AER_THRUST_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#endif
+
+#include <thrust/for_each.h>
+#include <thrust/complex.h>
+#include <thrust/inner_product.h>
+#include <thrust/transform.h>
+#include <thrust/transform_scan.h>
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <thrust/functional.h>
+#include <thrust/tuple.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/detail/vector_base.h>
+
+#ifdef AER_THRUST_CUDA
+#include <thrust/device_vector.h>
+#endif
+#include <thrust/host_vector.h>
+
+#include <thrust/system/omp/execution_policy.h>
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <complex>
+#include <cstdint>
+#include <string>
+#include <vector>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+
+#define AER_DEFAULT_MATRIX_BITS   6
+
+#define AER_CHUNK_BITS        21
+#define AER_MAX_BUFFERS       4
+#define AER_DUMMY_BUFFERS     4     //reserved storage for parameters
+
+
+#ifdef AER_THRUST_CUDA
+#define AERDeviceVector thrust::device_vector
+#else
+#define AERDeviceVector thrust::host_vector
+#endif
+#define AERHostVector thrust::host_vector
+
+#include "framework/utils.hpp"
+
+namespace AER {
+namespace QV {
+
+template <typename data_t> class Chunk;
+template <typename data_t> class DeviceChunkContainer;
+template <typename data_t> class HostChunkContainer;
+
+typedef struct _blocked_gate_params_ 
+{
+  uint_t mask_;
+  char gate_;
+  unsigned char qubit_;
+}BlockedGateParams;
+
+//========================================
+//  base class of gate functions
+//========================================
+template <typename data_t>
+class GateFuncBase
+{
+protected:
+  thrust::complex<data_t>* data_;   //pointer to state vector buffer
+  thrust::complex<double>* matrix_;   //storage for matrix on device
+  uint_t* params_;                    //storage for additional parameters on device
+  uint_t base_index_;               //start index of state vector 
+public:
+  GateFuncBase(void)
+  {
+    data_ = NULL;
+    base_index_ = 0;
+  }
+
```

Comment:
Including destructor in `GateBaseFunc` causes compiler error because it is called from device but it is defined as host code. 

`  virtual __host__ __device__ ~GateFuncBase(){}`
can be used and I could compile this, but actually there is no implementation that need destructor.

Reply:
I think we need to add this, maybe right now is not needed but maybe in the future a new derive class needs to access destructors through reference to base class. We avoid possible future bugs.

Reply:
I added